File: src/TaskOTime/TaskOTime.TimeTrackingServices/TimeItemBase.vb
Namespace ActiveDevelop.TimeTrackingServices
    ''' <summary>
    '''  bewaart een tijdregel met wijzigingsmeldingen, optionele tijdstippen en verwijzingen naar buren.
    '''  de regel beheert geen eigen sorteerlijst.  een omringende collectie kan de buurrelaties wijzigen.
    ''' </summary>
    ''' <remarks>
    '''  ontbrekende actievlaggen zijn niet zonder meer gelijk aan onwaar: zij kunnen een berekening
    '''  overslaan.  de collectieberekening van tijdsafstanden staat los van deze vlagafhankelijke logica.
    '''  eigenschapsmeldingen worden meteen doorgegeven en vormen geen gezamenlijke wijzigingstransactie.
    ''' </remarks>
    Public Class TimeItemBase
        Inherits ObservableObject
        Implements ITimeItem(Of Guid)

        Public Event EventTimeChanged As EventHandler
        Public Event ActionTargetInternalChanged As EventHandler

        Private _durationToNext As TimeSpan?
        Private _durationToPrevious As TimeSpan?
        Private _eventTime As DateTimeOffset?
        Private _idTimeItem As Guid
        Private _isStartAction As Boolean?
        Private _isEndAction As Boolean?
        Private _itemState As TimeItemState
        Private _lastChanged As DateTimeOffset?
        Private _isNotTimeItem As Boolean
        Private _processEndpointItem As ITimeItem(Of Guid)
        Private _previousItem As ITimeItem(Of Guid)
        Private _nextItem As ITimeItem(Of Guid)
        Private _dateCreated As DateTimeOffset?

        Public Property IDTimeItem As Guid Implements ITimeItem(Of Guid).IDTimeItem
            Get
                Return _idTimeItem
            End Get
            Set(value As Guid)
                SetProperty(_idTimeItem, value)
            End Set
        End Property

        Public Property DurationToNext As TimeSpan? Implements ITimeItem(Of Guid).DurationToNext
            Get
                Return _durationToNext
            End Get
            Set(value As TimeSpan?)
                SetProperty(_durationToNext, value)
            End Set
        End Property

        Public Property DurationToPrevious As TimeSpan? Implements ITimeItem(Of Guid).DurationToPrevious
            Get
                Return _durationToPrevious
            End Get
            Set(value As TimeSpan?)
                SetProperty(_durationToPrevious, value)
            End Set
        End Property

        ''' <summary>
        '''  meldt een gewijzigd tijdstip voordat de afzonderlijke tijdstipmelding en duurberekening volgen.
        ''' </summary>
        ''' <remarks>
        '''  bij een uitzondering wordt alleen het tijdstipveld teruggezet.  reeds uitgevoerde reacties
        '''  van afnemers worden niet teruggedraaid en er volgt geen extra herstelmelding.
        ''' </remarks>
        Public Property EventTime As DateTimeOffset? Implements ITimeItem(Of Guid).EventTime
            Get
                Return _eventTime
            End Get
            Set(value As DateTimeOffset?)
                Dim oldEventTime = _eventTime

                Try
                    If SetProperty(_eventTime, value) Then
                        OnEventTimeChanged()
                        CalculateDurationToLinkedItems()
                    End If
                Catch
                    _eventTime = oldEventTime
                    Throw
                End Try
            End Set
        End Property

        Protected Overridable Sub OnEventTimeChanged()
            RaiseEvent EventTimeChanged(Me, EventArgs.Empty)
        End Sub

        Protected Overridable Sub OnActionTargetInternalChanged()
            RaiseEvent ActionTargetInternalChanged(Me, EventArgs.Empty)
        End Sub

        Public Property IsStartAction As Boolean? Implements ITimeItem(Of Guid).IsStartAction
            Get
                Return _isStartAction
            End Get
            Set(value As Boolean?)
                SetProperty(_isStartAction, value)
                CalculateDurationToLinkedItems()
            End Set
        End Property

        Public Property IsEndAction As Boolean? Implements ITimeItem(Of Guid).IsEndAction
            Get
                Return _isEndAction
            End Get
            Set(value As Boolean?)
                SetProperty(_isEndAction, value)
                CalculateDurationToLinkedItems()
            End Set
        End Property

        Public Property ItemState As TimeItemState Implements ITimeItem(Of Guid).ItemState
            Get
                Return _itemState
            End Get
            Set(value As TimeItemState)
                SetProperty(_itemState, value)
            End Set
        End Property

        Public Property LastChanged As DateTimeOffset? Implements ITimeItem(Of Guid).LastChanged
            Get
                Return _lastChanged
            End Get
            Set(value As DateTimeOffset?)
                SetProperty(_lastChanged, value)
            End Set
        End Property

        Public Property DateCreated As DateTimeOffset? Implements ITimeItem(Of Guid).DateCreated
            Get
                Return _dateCreated
            End Get
            Set(value As DateTimeOffset?)
                SetProperty(_dateCreated, value)
            End Set
        End Property

        Public Property NextItem As ITimeItem(Of Guid) Implements ITimeItem(Of Guid).NextItem
            Get
                Return _nextItem
            End Get
            Set(value As ITimeItem(Of Guid))
                SetProperty(_nextItem, value)
                CalculateDurationToLinkedItems()
            End Set
        End Property

        Public Property PreviousItem As ITimeItem(Of Guid) Implements ITimeItem(Of Guid).PreviousItem
            Get
                Return _previousItem
            End Get
            Set(value As ITimeItem(Of Guid))
                SetProperty(_previousItem, value)
                CalculateDurationToLinkedItems()
            End Set
        End Property

        Public Property ProcessEndpointItem As ITimeItem(Of Guid) Implements ITimeItem(Of Guid).ProcessEndpointItem
            Get
                Return _processEndpointItem
            End Get
            Set(value As ITimeItem(Of Guid))
                SetProperty(_processEndpointItem, value)
            End Set
        End Property

        Public Property IsNotTimeItem As Boolean
            Get
                Return _isNotTimeItem
            End Get
            Set(value As Boolean)
                SetProperty(_isNotTimeItem, value)
            End Set
        End Property

        ''' <summary>
        '''  berekent per richting alleen wanneer de betreffende buur en de bijbehorende actievlag bestaan.
        ''' </summary>
        ''' <remarks>
        '''  binnen zo'n berekening wist een begin- of eindactie de duur, net als een ontbrekend tijdstip.
        '''  ontbreekt de buur of de benodigde vlag, dan blijft de bestaande duur in die richting staan.
        ''' </remarks>
        Protected Overridable Sub CalculateDurationToLinkedItems()
            If PreviousItem IsNot Nothing AndAlso IsStartAction.HasValue Then
                If IsStartAction.Value OrElse If(IsEndAction, False) OrElse Not EventTime.HasValue OrElse Not PreviousItem.EventTime.HasValue Then
                    DurationToPrevious = Nothing
                Else
                    DurationToPrevious = EventTime.Value - PreviousItem.EventTime.Value
                End If
            End If

            If NextItem IsNot Nothing AndAlso IsEndAction.HasValue Then
                If IsEndAction.Value OrElse If(IsStartAction, False) OrElse Not EventTime.HasValue OrElse Not NextItem.EventTime.HasValue Then
                    DurationToNext = Nothing
                Else
                    DurationToNext = NextItem.EventTime.Value - EventTime.Value
                End If
            End If
        End Sub

        Public Overrides Function ToString() As String
            Dim current = If(EventTime.HasValue, EventTime.Value.ToString("yy-MM-dd HH:mm"), "##-##-## ##:##")
            Dim previous = If(PreviousItem IsNot Nothing AndAlso PreviousItem.EventTime.HasValue, PreviousItem.EventTime.Value.ToString("yy-MM-dd HH:mm"), "##-##-## ##:##")
            Dim [next] = If(NextItem IsNot Nothing AndAlso NextItem.EventTime.HasValue, NextItem.EventTime.Value.ToString("yy-MM-dd HH:mm"), "##-##-## ##:##")

            Return $"--> {current} (<-- {previous} --> {[next]}"
        End Function
    End Class
End Namespace

File: src/TaskOTime/TaskOTime.TimeTrackingServices/TimeItemBase.cs
using System;

namespace ActiveDevelop.TimeTrackingServices
{
    /// <summary>
    /// Stores a time item with change notifications, nullable timestamps and references to its neighbors.
    /// The item does not maintain a sorted list; its containing collection may change the neighbor links.
    /// </summary>
    /// <remarks>
    /// Missing action flags are not equivalent to false: they can cause a calculation to be skipped.
    /// The collection's timestamp-distance calculation is separate from this flag-dependent logic.
    /// Property notifications are raised immediately and do not form an atomic change transaction.
    /// </remarks>
    public class TimeItemBase : ObservableObject, ITimeItem<Guid>
    {

        public event EventHandler EventTimeChanged;
        public event EventHandler ActionTargetInternalChanged;

        private TimeSpan? _durationToNext;
        private TimeSpan? _durationToPrevious;
        private DateTimeOffset? _eventTime;
        private Guid _idTimeItem;
        private bool? _isStartAction;
        private bool? _isEndAction;
        private TimeItemState _itemState;
        private DateTimeOffset? _lastChanged;
        private bool _isNotTimeItem;
        private ITimeItem<Guid> _processEndpointItem;
        private ITimeItem<Guid> _previousItem;
        private ITimeItem<Guid> _nextItem;
        private DateTimeOffset? _dateCreated;

        public Guid IDTimeItem
        {
            get
            {
                return _idTimeItem;
            }
            set
            {
                SetProperty(ref _idTimeItem, value);
            }
        }

        public TimeSpan? DurationToNext
        {
            get
            {
                return _durationToNext;
            }
            set
            {
                SetProperty(ref _durationToNext, value);
            }
        }

        public TimeSpan? DurationToPrevious
        {
            get
            {
                return _durationToPrevious;
            }
            set
            {
                SetProperty(ref _durationToPrevious, value);
            }
        }

        /// <summary>
        /// Raises the property change before the separate timestamp event and duration calculation.
        /// </summary>
        /// <remarks>
        /// If an exception occurs, only the timestamp field is restored. Subscriber side effects are
        /// not rolled back, and no additional notification announces the restoration.
        /// </remarks>
        public DateTimeOffset? EventTime
        {
            get
            {
                return _eventTime;
            }
            set
            {
                var oldEventTime = _eventTime;

                try
                {
                    if (SetProperty(ref _eventTime, value))
                    {
                        OnEventTimeChanged();
                        CalculateDurationToLinkedItems();
                    }
                }
                catch
                {
                    _eventTime = oldEventTime;
                    throw;
                }
            }
        }

        protected virtual void OnEventTimeChanged()
        {
            EventTimeChanged?.Invoke(this, EventArgs.Empty);
        }

        protected virtual void OnActionTargetInternalChanged()
        {
            ActionTargetInternalChanged?.Invoke(this, EventArgs.Empty);
        }

        public bool? IsStartAction
        {
            get
            {
                return _isStartAction;
            }
            set
            {
                SetProperty(ref _isStartAction, value);
                CalculateDurationToLinkedItems();
            }
        }

        public bool? IsEndAction
        {
            get
            {
                return _isEndAction;
            }
            set
            {
                SetProperty(ref _isEndAction, value);
                CalculateDurationToLinkedItems();
            }
        }

        public TimeItemState ItemState
        {
            get
            {
                return _itemState;
            }
            set
            {
                SetProperty(ref _itemState, value);
            }
        }

        public DateTimeOffset? LastChanged
        {
            get
            {
                return _lastChanged;
            }
            set
            {
                SetProperty(ref _lastChanged, value);
            }
        }

        public DateTimeOffset? DateCreated
        {
            get
            {
                return _dateCreated;
            }
            set
            {
                SetProperty(ref _dateCreated, value);
            }
        }

        public ITimeItem<Guid> NextItem
        {
            get
            {
                return _nextItem;
            }
            set
            {
                SetProperty(ref _nextItem, value);
                CalculateDurationToLinkedItems();
            }
        }

        public ITimeItem<Guid> PreviousItem
        {
            get
            {
                return _previousItem;
            }
            set
            {
                SetProperty(ref _previousItem, value);
                CalculateDurationToLinkedItems();
            }
        }

        public ITimeItem<Guid> ProcessEndpointItem
        {
            get
            {
                return _processEndpointItem;
            }
            set
            {
                SetProperty(ref _processEndpointItem, value);
            }
        }

        public bool IsNotTimeItem
        {
            get
            {
                return _isNotTimeItem;
            }
            set
            {
                SetProperty(ref _isNotTimeItem, value);
            }
        }

        /// <summary>
        /// Calculates each direction only when its neighbor and the corresponding action flag exist.
        /// </summary>
        /// <remarks>
        /// Within either calculation, a start or end action clears the duration, as does a missing timestamp.
        /// If the neighbor or required flag is absent, the existing duration in that direction is retained.
        /// </remarks>
        protected virtual void CalculateDurationToLinkedItems()
        {
            if (PreviousItem is not null && IsStartAction.HasValue)
            {
                if (IsStartAction.Value || (IsEndAction ?? false) || !EventTime.HasValue || !PreviousItem.EventTime.HasValue)
                {
                    DurationToPrevious = default;
                }
                else
                {
                    DurationToPrevious = EventTime.Value - PreviousItem.EventTime.Value;
                }
            }

            if (NextItem is not null && IsEndAction.HasValue)
            {
                if (IsEndAction.Value || (IsStartAction ?? false) || !EventTime.HasValue || !NextItem.EventTime.HasValue)
                {
                    DurationToNext = default;
                }
                else
                {
                    DurationToNext = NextItem.EventTime.Value - EventTime.Value;
                }
            }
        }

        public override string ToString()
        {
            string current = EventTime.HasValue ? EventTime.Value.ToString("yy-MM-dd HH:mm") : "##-##-## ##:##";
            string previous = PreviousItem is not null && PreviousItem.EventTime.HasValue ? PreviousItem.EventTime.Value.ToString("yy-MM-dd HH:mm") : "##-##-## ##:##";
            string next = NextItem is not null && NextItem.EventTime.HasValue ? NextItem.EventTime.Value.ToString("yy-MM-dd HH:mm") : "##-##-## ##:##";

            return $"--> {current} (<-- {previous} --> {next}";
        }
    }
}